hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastComparisonPredicateUtils.java

```diff
@@ -122,11 +122,14 @@ private static boolean checkOperandTypesWithPrecedence(
 
         if (highHZType.getTypeFamily() != lowHZType.getTypeFamily()
             && !(highHZType.getTypeFamily().isNumeric() && lowHZType.getTypeFamily().isNumeric())) {
-            // Types cannot be converted to each other, throw.
-            if (throwOnFailure) {
-                throw callBinding.newValidationSignatureError();
-            } else {
-                return false;
+
+            if (bothNotTemporal(highHZType, lowHZType) || nonConvertibleTemporalTypes(highHZType, lowHZType)) {
```

Comment:
`lowHZType.getConverter().canConvertTo(highHZType.getTypeFamily())` should be sufficient to cover all cases.

Reply:
I assume that you meant `!lowHZType.getConverter().canConvertTo(highHZType.getTypeFamily())` (see `!` in the beginning of the expression)

I ran `ComparisonPredicateIntegrationTest` locally and some tests failed. For example [`testString` on the line with `checkUnsupportedColumnColumn(ExpressionTypes.STRING, ExpressionTypes.allExcept(ExpressionTypes.STRING, ExpressionTypes.CHARACTER));`](https://github.com/hazelcast/hazelcast/blob/master/hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/ComparisonPredicateIntegrationTest.java#L101)

I added more descriptive error message so it fails with in this case
`java.lang.AssertionError: Must fail for operands with types String and Boolean`. 
My understanding right now we have an issue with either in tests that test unsupported conversions or in `Converter.canConvertTo`

Reply:
I see. Then we may use `canConvertTo` only for temporal types. This is a more general solution than try to encode `DATE`/`TIME` incompatibility manually.